include/toolchain/gcc.h

```diff
@@ -88,7 +88,7 @@
 /* The GNU assembler for Cortex-M3 uses # for immediate values, not
  * comments, so the @nobits# trick does not work.
  */
-#if defined(CONFIG_ARM)
+#if defined(CONFIG_ARM) || defined(CONFIG_ARM64)
```

Comment:
Yes, that was the other option on the table but my idea was that ARM and ARM64 are different enough to deserve a new separate architecture (not even the assembly language is the same when using AArch64).

Reply:
it isn't for x86_64 either, we have separate .S files
see how we split arch/x86/core/ia32 and arch/x86/core/intel64
This is WIP but we're hoping to get most of the C code the same between the two.
I don't know if this is feasible for ARM though, just a passing comment.

Reply:
@andrewboie (and maybe @stephanosio and @bbolen) can we discuss this better? I'm still against merging ARM and ARM64 because they are different architectures with very little in common. Maybe some C code (and Kconfig symbols) could be shared and reuse but I wonder if it is really worth it.

I tried to merge the two while reworking this commit but it really comes down to a lot of boilerplate code to be able to distinguish between ARM and ARM64 and including / using the correct code specific to each architecture. Also the cmake extensions must be instructed to differentiate between the two.

Linux (or other RTOS like LK) are treating ARM and ARM64 as two different architectures so I really wonder if we want to do the same.

Reply:
@carlocaione At least, the general trend in Zephyr is, regardless of right or wrong, to have the same root arch for 32-bit and 64-bit variants.

For example, both x86 and RISC-V use the same root arch directory name as well as Kconfig symbol (`X86` and `RISCV`, respectively) for both 32-bit and 64-bit implementations.

In terms of Kconfig symbols:
* For RISCV, the cross-arch symbol `64BIT` is used to indicate that the RISCV64 arch is used.
* For x86, an additional symbol, `X86_64`, is set alongside `X86` and `64BIT` to indicate that the AMD64 arch is used (`X86_64` could be thought of as a helper symbol implying both `X86` and `64BIT`, though this needs to be consolidated at the time of writing).

While I think this could require a further, much more in-depth, discussion, generally speaking, given that there is really no absolute/correct answer to this, it would make the most sense to follow what other archs are already doing for the sake of consistency.

Some arguments in favour of having single root arch for ARM are as follows:
* While there are some major fundamental differences between AArch32 and AArch64, this does not change the fact that they are both ARM architectures and implement a fairly similar set of instructions as well as architecture features.
* Generally speaking, it would be easier to mix and glue any common code between AArch32 and AArch64 if they are under the same arch.
* Some common architecture features, such as TrustZone, could take advantage of being under the same logical arch (at least, the initial designation as such would encourage more consistent 
 and commonly shared implementation across the AArch32 and AArch64).
* If we ever need to implement switching between AArch32 and AArch64 modes in the future, it would be easier to do this with single root arch than two apparently independent arch definitions.

In terms of single ARM root arch implementation, we could probably do something like below:
* For Kconfig:
  * Set `ARM` for both AArch32 and AArch64 (indicating we are on "ARM-family" arch).
  * Select `64BIT` for AArch64, and use `ARM64` (set when ARM && 64BIT) as a helper symbol for quick reference.
* For directory structure:
  * arch/arm/core: Common config and code across AArch32 and AArch64 for all Cortex variants go in here
  * arch/arm/core/cortex_m: Move the existing Cortex-M-specific code here (both ARMv7 and ARMv8).
  * arch/arm/core/aarch32: Move the existing AArch32 config and code into this directory. Also make this the default directory for both Cortex-A and Cortex-R as they practically have the same base arch (really only Cortex-M is the special one here, the only major difference between Cortex-A and Cortex-R is MMU vs. MPU; consider adding aarch32/mmu and aarch32/mpu in the future for addressing that, any other AArch32 Cortex-A/-R-specific quirks should be manageable with `#if`).
  * arch/arm/core/aarch64: Add new AArch64 config and code to this directory. No Cortex-A sub-directory as there is really no need (pretty much the same idea as the aarch32 directory).

@ioannisg what do you think about this?

Reply:
When reworking this I was doing something similar:
- `arch/arm/core` common code for AArch32 and AArch64
- `arch/arm/core/32/cortex_{m,r}` since ARMv8-M is actually AArch32 (why do you want to move that to the common core directory?)
- `arch/arm/core/64/cortex_a` for consistency

This is mirrored into `include/arch/arm/{32,64}`

Reply:
> since ARMv8-M is actually AArch32 (why do you want to move that to the common core directory?)

The rationale is something like below:

* arch/arm/core: ARM common
* arch/arm/aarch32: AArch32 Cortex-A and Cortex-R
* arch/arm/aarch64: AArch64 Cortex-A (at the moment, there is really nothing else and will probably stay this way)
* arch/arm/cortex_m: (AArch32) Cortex-M

The reason for having Cortex-M as a separate directory under `arch/arm/core` is that, while it implements AArch32 instruction sets,  it really is a different architecture from the normal "AArch32" architectures (ARMv*-A, ARMv*-R); i.e. it has its own CPU vector, reset code, interrupt/exception handling mechanism, interrupt controller, context switching and many more ... (at the moment, I am actually thinking about splitting Cortex-R from Cortex-M completely because there is really very little in common with Cortex-M).

Reply:
@stephanosio I do not particularly like this solution because keeping the code organized in this way is confusing.

For example much of the code now shipped in `include/arch/arm` is shared between Cortex-R and Cortex-M, where should that be located according to this scheme? We could keep that in `include/arch/arm/` but then that has nothing to do with the subdirectory `aarch64` in `include/arch/arm/aarch64`. Otherwise we could move that into `include/arch/arm/aarch32/` but that is even more confusing with the Cortex-M code split between `include/arch/arm/aarch32` and `include/arch/arm/cortex-m`.

The same reasoning is valid for the code in `arch/arm/core`.

To be able to fully use your proposal we should separate Cortex-M code from Cortex-R and that's basically another story.

So my proposal is:
- `arch/arm/core`: ARM common
- `arch/arm/aarch32`: AARch32 Cortex-A, Cortex-R and Cortex-R/M (that is current code shared between Cortex-R and Cortex-M)
- `arch/arm/aarch32/cortex-m`: (AArch32) Cortex-M
- `arch/arm/aarch64`: (AArch64) Cortex-A

Reply:
> @andrewboie (and maybe @stephanosio and @bbolen) can we discuss this better? I'm still against merging ARM and ARM64 because they are different architectures with very little in common.

As I said before, this was a passing observation. I don't really care if you merge it with the other ARM targets or not. You should really discuss this with the other ARM maintainers (cc @ioannisg )

Reply:
@carlocaione I meant to say `arch/arm/core/aarch{32,64}`, not `arch/arm/aarch{32,64}`. I apologise for causing any confusion in regards to this.

All the details can be worked out when we write up an RFC for this in the future; but in general, my thought on this is as follows (for future reference when we start working on RFC):

* I am not totally against `arch/arm/core/aarch32/cortex_m`, as this, depending on how much is shared between Cortex-M and Cortex-R/Cortex-A AArch32, could be a more logical path to take. This requires a further in-depth assessment.
* I believe we should organise Cortex-R and AArch32 Cortex-A together, and only handle the differences between them using conditional compilation as they are practically the same arch (other than MPU vs MMU).
* It looks as though Cortex-R shares a lot of code with Cortex-M at the moment because 1) in some parts, it was deliberately written to do that even while making some less-than-ideal design choices, 2) most of the seemingly "shared" code is not actually used by Cortex-R port- note that the Cortex-R port is quite incomplete at this time, 3) a lot of code is placed into the same file with ifdefs, when they should really be separate files (at the moment, big chunks of Cortex-M and Cortex-R-specific code coexist in the same file; even for many small chunks, if you take a careful look into the code, you will see that they are not really "shared", but conditionally selected for either Cortex-M or Cortex-R).

I am working on various improvements to the Cortex-R port at the moment and, as these get merged, Cortex-R port will deviate significantly from Cortex-M port to the point that it no longer really makes sense to keep them together (mainly because they are, in fact, two fundamentally different archs).

Since we do not have a very clear picture as to how the coexistence among Cortex-M, Cortex-R, Cortex-A AArch32 and Cortex-A AArch64 will work out, I think it would be better to just proceed with what you have right now for this PR and do a major refactoring in the future (with a preceding RFC for in-depth discussion of course).

By the way, after I wrap things up with Cortex-R, I am also planning to add basic ARMv7 Cortex-A (i.e. AArch32) support (mainly for TI Sitara processors); so after that, we should have a reasonably clear idea as to how all different Cortex series architectures should coexist).